### ISSUE TITLE
General usage questions

### ISSUE THREAD

**Issue by alexcosmin** [deepfakes/faceswap-playground#184]
## Environment information
- **Operating system and version:** macOS High Sierra 10.13.5
- **Processor:** 2.2 GHz Intel Core i7
- **Memory:** 16 GB 1600 MHz DDR3
- **VRAM:** 1536 MB
- **GPU:** Intel Iris Pro, built in 
- **Python version:** 3.5.0
- **Faceswap version:** 5275718365847f85722f3c39a429aac2a42bd413
- **Faceswap method:** CPU
- **Running with Docker:** No
- **CUDA support:** No
---
I cloned the repo and followed the [installation](https://github.com/deepfakes/faceswap/blob/master/INSTALL.md) and [usage](https://github.com/deepfakes/faceswap/blob/master/USAGE.md) instructions.   
Running from the command line (default options), I'm currently using the [training data provided](https://anonfile.com/p7w3m0d5be/face-swap.zip) in order to train the model (using the default trainer which seems to be **Original**) for the 2 faces (trump and cage). 
I checked the full lists of arguments for each of the commands:
```
python faceswap.py extract -h
python faceswap.py train -h
python faceswap.py convert -h
```
So far so good. I'd like to improve my understanding of the whole process, therefore I have a couple of questions:
1. **Why don't the losses have strictly descending values?** (as I understand, each output line is printed after each 100 iterations). For example:
```
saved model weights loss_A: 0.05456, loss_B: 0.04947
saved model weights loss_A: 0.05491, loss_B: 0.05097
```
[This](https://imgur.com/a/mo3dlZu) is how my current progress looks like.  

2. From my [current progress](https://imgur.com/a/mo3dlZu) I understand that I'm at iteration number ~~40704~~ 4074 (currently running for approx. 28 hours). I'm training on 319 images of Cage and 377 images of Trump. **What is the batch size?** (couldn't find this in the code) According to [this blogpost](https://medium.com/@sagarsharma4244/epoch-vs-iterations-vs-batch-size-4dfb9c7ce9c9) I should be able to calculate the **epochs number** I'm currently at. (which is actually what I want to know)

3. Can I stop the training command that's currently running and restart it with the same command?
```
python faceswap.py train -A data/trump -B data/cage -m models/ -p
```
**Will it resume from the current state of the trained model?** (I'm assuming the progress is saved in the `models/*.h5` files) [This issue](https://github.com/deepfakes/faceswap-playground/issues/60) suggests the answer was yes in February, but is this still the case?

4. Can I just open a new shell (while the *train* command from above is still running) and perform the *convert* command? 
```
python faceswap.py convert -i photo/trump/ -o output/ -m models_test/
```
Before running this, I would simply copy+paste the `models/*.h5` files to `models_test`. The idea here would be to periodically check (out of curiosity) what would the output with the current trained model. 

5. I'm running the *train* command with the preview feature. **I'm trying to understand what does the preview feature show**. It started from [this](https://imgur.com/a/UIO0uNd) and now it shows [this](https://imgur.com/a/mspRco4). Is it correct that the less blurry the images become in the preview, the better the output will be? Is it the preview of the output? Meaning that if I stop the training now and perform the *convert* function, I'll get the results shown in the preview? (currently the blurry images) I understand from [this comment](https://github.com/deepfakes/faceswap/issues/211#issuecomment-366843319) that when the losses achieve values of < 0.01, the results will be decent. 

Any answer would be appreciated. I imagine there are multiple people with these kind of questions, therefore maybe we can start making a FAQ section?

**Comment by bryanlyon** [comment]
Loss is not a linear scale, given different images you will get different results.  Each iteration had a different sunset of images so loss will vary.  It should generally tend downward.  It's gradient descent but will never be always down every iteration.

Default batch size is 64.  You're definitely over training your model.  It needs many more images for that level of training.  Generally keep it your iterations between 10-30x your images.

Training status is saved and will resume from that state (but iteration count will reset, so keep track of each run)

You shouldn't run convert whole training, they'll compete for your GPU and possibly corrupt your model file due to read write contention.

The preview shows the original image, regenerating that same face using the a decoder (used to generate the loss) and the other face (to show how close it is to a good match).

**Comment by alexcosmin** [comment]
Thank you for answering my questions @bryanlyon ! Much appreciated. 
I'd like to ask for a few clarifications:

2. I think I asked the wrong question regarding the batch size. Seems like I need to know the **number of batches** (according to [this](https://towardsdatascience.com/epoch-vs-iterations-vs-batch-size-4dfb9c7ce9c9)), in order to calculate the epochs number I'm currently at. What would this be?  I'm training on 319 images of Cage and 377 images of Trump, therefore my understanding is that I have 696 training examples. 

> You're definitely over training your model. It needs many more images for that level of training. 

I made a mistake when I posted the question, **I was actually at iteration number 4074 and not 40704**.
Am I overfitting my model? I thought the idea was to train it until the losses achieve values of < 0.01 and/or the preview tool shows images that are not blurry (as you can see [my current preview](https://imgur.com/a/mspRco4) shows images that are still blurry).

> Generally keep it your iterations between 10-30x your images.

How did you come up with this interval of 10-30x? Trial and error or theory? 
So with 696 training examples, I should not exceed 696*30 = 20880 iterations. Is that correct?

Additional question:

6. Is it correct that the **batch_size** somewhat corresponds to the **output image size** (seems to be 64x64)? Or just a coincidence? Why is the batch_size chosen to be 64? What happens if I increase it to 128, in terms of performance?

**Comment by samsonsite1** [comment]
I'm just guessing batch size relates to the number of images worked on at the same time, but I don't know how that tranlsates into how much system memory or GPU memory being used.

I also wondered how does the program pick which images to show in the training window, but I'm guessing the images are not random, but sequentially picked over some range.

**Comment by alexcosmin** [comment]
@samsonsite1 
According to [this](https://towardsdatascience.com/epoch-vs-iterations-vs-batch-size-4dfb9c7ce9c9), the **batch size** is the *total number of training examples present in a single batch*. And a **batch** is the *amount of images passed into the neural network at once*. 

> I also wondered how does the program pick which images to show in the training window, but I'm guessing the images are not random, but sequentially picked over some range.

Good question, would be good to know.

**Comment by alexcosmin** [comment]
If anyone is curious, those are my results in the following conditions:

Started training model from scratch, running with the default options, using the [provided training data](https://anonfile.com/p7w3m0d5be/face-swap.zip).
**Training data**: 319 pictures of Nicolas Cage and 377 pictures of Donald Trump
**Model used**: Model_Original
**Model trained for**: 43 hours, 6418 iterations, batch size of 64 images
**Hardware used**: MacBook Pro (Mid 2015), CPU with 8 logical cores, integrated GPU with 1536 MB VRAM
**Started with**: loss_A: 0.19021, loss_B: 0.18308
**Ended with**: loss_A: 0.04791, loss_B: 0.04353
**Results**: 318 swaps. The results are not photorealistic, due to the loss values not achieving low enough values (< 0.01) during the training of the neural network. In other words, the neural network needs a longer training in order to achieve better results and also more training data as input. 
**Failure reasons in facial recognition**: The facial recognition implementation used successfully detected faces in 318 images out of 377 images. It seems like most of the failures occur due to pose issues and accessories used in the picture (for example a hat).

Few examples of the output (randomly chosen):
![3146925](https://user-images.githubusercontent.com/5630100/41609419-ffc7ded4-73ea-11e8-85c1-2c9f847aef3c.jpg)
![51130785](https://user-images.githubusercontent.com/5630100/41609432-05b35b20-73eb-11e8-8cf1-b8d767175fbc.jpg)
![51834796](https://user-images.githubusercontent.com/5630100/41609442-0a63443c-73eb-11e8-87c9-630b3583edb8.jpg)
![55871042](https://user-images.githubusercontent.com/5630100/41609455-140af160-73eb-11e8-8acd-0a33bd618ef6.jpg)
![112270915](https://user-images.githubusercontent.com/5630100/41609467-1bf36574-73eb-11e8-8d5c-a726b32cf446.jpg)

**Comment by nathanshipley** [comment]
Hi - just to note, it's going to go **very** slow without an Nvidia GPU in your system.  To the point where I consider the tool almost unusable if you're only using CPU.  As a comparison, I ran training on a batch of images last night for about 12 hours using OriginalHighRes (which takes a lot longer than Original), batch size 64, on a system with 2 1080Ti cards in it and had 79000 epochs when I came in in the morning.  The results look great.

Basically, you need a legit video card.  Beyond that, I've used the software with old cheesegrater Mac Pro towers that have Nvidia cards in them, but it's quite an ordeal to get Tensorflow working in GPU mode since Google doesn't support it.  You end up needing to compile TF from source.

Much easier to work on a Linux rig or use an AWS P2 instance that has a GPU in it.  Good luck!

**Comment by alexcosmin** [comment]
@nathanshipley 
Appreciate the message, thank you for sharing your experience! 
Yes, I'm aware of the fact that it'll be super slow if I'm only using a CPU. Unfortunately I don't have access to a better computer right now, but that's okay, since at this point I'm mainly interested to understand how the whole thing works, what are the possible configurations.

> had 79000 epochs when I came in in the morning

You mean iterations? Just making sure I'm getting the terms correctly.

**Comment by alexcosmin** [comment]
@Kirin-kun Thanks for clarifying that! Iterations it is. 

@torzdf Could be an idea to rename the -ep parameter. I can do it in the weekend, if you're not faster than me.

**Comment by agilebean** [comment]
@alexcosmin and @torzdf : 
yes please rename! i was really confused...
you would make a meaningful contribution even though it's not challenging.

**Comment by torzdf** [comment]
It's been renames in latest commit

**Comment by alexcosmin** [comment]
Thank you for all the responses! I'll close this now.
